server/src/main/java/org/elasticsearch/snapshots/SnapshotInfo.java

```diff
@@ -428,6 +448,18 @@ public Version version() {
         return version;
     }
 
+    /**
+     * Returns the custom metadata that was attached to this snapshot at creation time.
+     * @return custom metadata
+     */
+    @Nullable
+    public Map<String, Object> userMetadata() {
+        if (userMetadata != null) {
+            return Collections.unmodifiableMap(userMetadata);
```

Comment:
No need to wrap the map here imo. I think the caller should make sure to not pass a mutable map here (or mutate the map after passing it here). If we're going for potentially nested maps and lists the code that creates the metadata map should make sure it passes a structure that's immutable instead I think to prevent all modifications.

Reply:
It's a good point that the inner objects may be mutable even if we prevent this map from being mutable. This was mostly out of habit, but you're right that in this case it doesn't buy much unless we want to go to a lot more effort to dig in and make the whole thing immutable, which probably isn't worth it.